README.md
# content-change-ledger

Record why a content field changed and connect it to a release or owner.

- **Repository:** [edilec/content-change-ledger](https://github.com/edilec/content-change-ledger)
- **Area:** Content & Publishing
- **License:** MIT

## The problem

"Who changed the pricing page, why, and which release shipped it?" is usually
answered by asking three people. The commit says what the bytes became, the
ticket says what someone intended a fortnight earlier, and the release notes say
neither.

A ledger answers it in one file. Each change appends one event: the content hash
before, the content hash after, the reason, the owner, the release id, and the
hash of the event before it. Because every event's hash covers the event before
it, an event that is quietly edited or lifted out of the middle of the file
stops matching, and `verify` says exactly which line broke.

Nothing is ever edited. A mistake in a recorded event is fixed by appending a
**correction** that names it -- the original stays in the file, byte for byte.

## Install

Node 22 or newer. No runtime dependencies and no build step.

```sh
npm install content-change-ledger
```

Or run it from a checkout:

```sh
node bin/content-change-ledger.mjs --help
```

## Commands

```sh
# record a change
content-change-ledger append ledger.jsonl \
  --subject content/pricing.md --action update \
  --after-file content/pricing.md \
  --reason "Team plan moved to 49 USD" --owner content-team --release 2026.09.1

# fix a mistake in an earlier event without touching it
content-change-ledger append ledger.jsonl \
  --subject content/pricing.md --action correct --corrects evt-0002 \
  --after sha256:d67ce779... \
  --reason "evt-0002 recorded the wrong release" --owner content-team --release 2026.09.2

# check the whole chain against a policy and a head checkpoint
content-change-ledger verify examples/ledger.jsonl --config examples/ledger-policy.json

# read the history back
content-change-ledger history examples/ledger.jsonl --subject content/pricing.md
content-change-ledger history examples/ledger.jsonl --release 2026.09.1 --json
```

| Option | Verb | Meaning |
| :--------------- | :------ | :--------------------------------------------- |
| `--config FILE` | verify | Policy and head checkpoint (JSON) |
| `--root DIR` | all | Confine ledger paths to this directory (default: cwd) |
| `--json` | all | Emit the machine-readable report on stdout |
| `--subject S` | history | Only events about this subject |
| `--owner O` | history | Only events recorded by this owner |
| `--release ID` | history | Only events in this release |
| `--action A` | history | Only `create`, `update`, `delete` or `correct` |
| `--since` / `--until` | history | Bound the answer by `recordedAt`, written `2026-09-13T09:30:00.000Z` |
| `--limit N` | history | Bound the whole answer; a cut answer is `incomplete` |
| `--subject`, `--action`, `--reason`, `--owner`, `--release` | append | Required |
| `--after HASH` / `--after-file FILE` | append | The content hash after the change |
| `--before HASH` | append | Override the inherited before hash |
| `--id ID` | append | Event id (default: `evt-NNNN` by position) |
| `--corrects ID` | append | The earlier event this correction refers to |
| `--recorded-at TS` | append | UTC timestamp (default: now) |

Repository scripts: `npm run lint`, `npm test`, `npm run test:coverage`,
`npm run example`, `npm run pack:check`, and `npm run check` for all of them.
The deliberately tampered ledger is checked with:

```sh
node bin/content-change-ledger.mjs verify examples/broken-ledger.jsonl
```

It reports the chain break, the subject discontinuity and the reused event id,
and exits 1.

## The event

One JSON object per line, twelve fields, all strings or null:

```json
{"id":"evt-0002","subject":"content/pricing.md","action":"update",
 "beforeHash":"sha256:197c095e...","afterHash":"sha256:d67ce779...",
 "reason":"Team plan moved to 49 USD and seat counts were added",
 "owner":"content-team","releaseId":"2026.09.1","recordedAt":"2026-09-04T11:15:00.000Z",
 "corrects":null,"previousHash":"sha256:3baa98d2...","hash":"sha256:d1b29e5a..."}
```

`hash` is the SHA-256 of the canonical JSON of the other eleven fields, in a key
order fixed in code. `previousHash` is one of them, which is what chains the
file together. The full format, the action rules and the rule catalog are in
[docs/ledger-rules.md](./docs/ledger-rules.md).

## Outputs

`--json` writes the report to stdout, and only the report, so it can be piped
straight into a parser. Diagnostics always go to stderr.

```json
{
  "schemaVersion": "1",
  "tool": "content-change-ledger",
  "status": "fail",
  "summary": { "checked": 4, "errors": 3, "warnings": 0, "info": 1, "events": 4,
               "corrections": 1, "files": 1, "filesRead": 1, "subjects": 2 },
  "findings": [
    {
      "ruleId": "chain-previous-hash-mismatch",
      "severity": "error",
      "message": "This event follows line 1 in the file but records a different previousHash, so an event between them was removed or replaced.",
      "location": { "file": "examples/broken-ledger.jsonl", "pointer": "/events/1" },
      "line": 2,
      "evidence": "recorded sha256:d1b29e5a..., expected sha256:3baa98d2...",
      "suggestion": "Restore the missing event; the chain is only continuous when each event names the hash before it."
    }
  ]
}
```

`history --json` returns the same envelope with an `entries` array. Each entry
is a copy of the stored event plus `supersededBy`, the ids of later corrections
that name it, and the ledger `file` it came from. That annotation is computed
for the reader and is never written into the file.

Every untrusted string in either output -- an id, a path, a message, an excerpt
-- is bounded and escaped. Control characters (C0, DEL and C1, which is where
U+0085 hides), line separators, and bidirectional overrides become visible
escape text. File labels additionally double literal backslashes, so an actual
control character and text that merely spells its escape cannot produce the
same label. A long file label carries a short prefix, its UTF-16 length, and a
SHA-256 digest of its full UTF-16 path. A crafted id or file name cannot forge
a line in the human report or reverse what it says.

A line or a configuration file that does not parse is reported by position,
line and column -- never by quoting it back. `JSON.parse` embeds the input in
one of its two error messages, so a file short enough to be only a credential
would otherwise be reproduced in full by its own failure, and escaping would
not stop it: the quotation sits at the front of the message while the bound
cuts from the back.

## Exit codes

| Code | Meaning |
| ---: | :------------------------------------------------------------- |
| `0` | The ledger verified, or the query was answered in full |
| `1` | The ledger failed verification |
| `2` | Invalid usage or configuration, or evidence not read in full |

On exit 2, stdout is **empty** for a usage or configuration error -- the run
never had a subject to report on -- and carries a report with status
`incomplete` when a named input could not be read, so a consumer knows which
one. A consumer that pipes stdout must handle both.

## Limits and non-goals

**This is not tamper-proofing, and the README will not pretend otherwise.**

- **A truncated tail is undetectable on its own.** Deleting the last events
  leaves a perfectly consistent chain. Only a head checkpoint held somewhere the
  ledger's writer cannot reach (`head` in the configuration) detects it. Without
  one, every report carries the `tail-not-anchored` note rather than implying the
  end of the file was checked.
- **Anyone who can rewrite the whole file can forge a consistent one.** The chain
  detects an event altered or removed *without* recomputing everything after it.
  It proves nothing against a writer who rebuilds the ledger from the first line.
  There are no signatures and no external anchor.
- **The timestamps are claims, not evidence.** `recordedAt` is whatever the
  caller supplied, and nothing corroborates it. The library never reads a clock:
  the command line is the only part of the tool that does, and only to default
  `--recorded-at` to this process's wall clock, which is a claim by the machine
  that ran the append and nothing more.
- **The content hashes are claims too.** `verify` checks the ledger against
  itself; it does not open your content files and re-hash them, so it cannot
  tell you whether `content/pricing.md` still matches the last `afterHash`
  recorded for it. `--after-file` hashes real bytes, but only at the moment of
  the append.
- **It cannot judge a reason or an owner.** `reason`, `owner` and `releaseId`
  are free text. A configured `owners` list and `releaseIdPattern` check spelling,
  not accountability.
- **One writer at a time.** There is no lock. Two processes appending
  simultaneously can both chain to the same event; the fork is detectable at the
  next `verify` as `chain-previous-hash-mismatch`, but it is not prevented.
- **A correction does not change what the earlier event says.** Any consumer that
  reads one event in isolation will still read the original. Read the history.
- **No network, ever.** No remote timestamping, no attestation service, no
  telemetry. Evidence the tool did not obtain locally is reported as unverified,
  and unverified is never a pass.
- **It is a file, not a database.** Every verb reads the whole ledger; the bounds
  in [docs/ledger-rules.md](./docs/ledger-rules.md) say how far that goes before
  the run reports `incomplete` instead of truncating. The one bound that is not
  a finding is `maxFiles`: naming more ledgers than it allows is refused before
  any of them is read, so it is a usage error with empty stdout rather than a
  report.

## License

MIT. See [LICENSE](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project are documented in this file.

## Unreleased

### Fixed

- a ledger line or a `--config` file that does not parse is no longer quoted
  back. `JSON.parse` embeds the input in one of its two error messages
  (`Unexpected token 'A', "AKIA..." is not valid JSON`), so a file short enough
  to be only a credential was reproduced in full by the `event-not-json`
  finding on stdout and by `Configuration is not valid JSON:` on stderr. Worse
  for a long line: V8 quotes a window around the offending character wherever
  it sits, so the message could show bytes from past the 120-character bound
  the `evidence` excerpt stops at. `parseFailureDetail` in `src/event.mjs` now
  keeps the position, line and column and drops the quotation;

### Added

- an append-only JSON Lines ledger of content change events: each event records
  the before and after content hashes, the reason, the owner, the release id,
  the UTC instant it was recorded, and the hash of the event before it;
- a hash chain whose canonical form has a key order fixed in code, so the same
  event hashes identically on any machine, and an event edited or lifted out of
  the middle of a file is detectable at the exact line;
- corrections as append-only events: `appendEvent` never rewrites a line, the
  event a correction names keeps its bytes, and `history` annotates the
  corrected event with `supersededBy` for the reader without storing it;
- verification of duplicate event ids, self-recorded hashes, chain continuity,
  the genesis event, per-subject before/after continuity, the subject state
  machine, correction targets, and an optional head checkpoint;
- `history` queries by subject, owner, release, action and time window, where
  every filter is validated -- a malformed window is refused, not answered --
  and an answer cut short by the limit on the whole answer is `incomplete`
  rather than quietly short;
- explicit bounds on bytes, events, event size, field length, query results and
  wall-clock time, each reported as `incomplete` with a finding naming the
  limit, plus `maxFiles`, which bounds the invocation and is refused as a usage
  error before any evidence is read;
- a CLI with `verify`, `history`, `append`, `--root`, `--json` and exit codes
  0 / 1 / 2;
- a clean example ledger with a policy file, and a deliberately tampered one;
- the format, rule catalog, limits and determinism guarantee in
  `docs/ledger-rules.md`.

### Defended

Each of these is a guarantee with a test that fails when the guarantee is
removed from the source:

- evidence that was not obtained is never a verdict. After a line that could not
  be parsed, the subject state machine, the before/after continuity check, the
  correction-target check and the chain link across the gap all stop reporting,
  and the run is `incomplete`;
- `pass` with `checked: 0` cannot be emitted: an empty ledger is `ledger-empty`
  and `incomplete`, and the report builder refuses a pass that verified nothing.
  Every rule that was observed to make a real run `incomplete` is also refused on
  a report that is not, so a rule quietly dropped from that list fails a test;
- severity comes from one frozen `ruleId -> severity` table, and what each rule
  does to a run is pinned by a real run: every rule in the catalog has a fixture
  driven through the command line, asserting the exit code, the status and the
  error, warning and info counts it produces. A severity changed in the table,
  the documented catalog and a test map together still fails, because the
  observed outcome changes. An unknown rule id throws, and the lookup that
  throws is exported so that guard can be called;
- ledgers and the configuration file are both decoded with
  `TextDecoder('utf-8', { fatal: true })`; encoding validity is never inferred
  from decoded text;
- path confinement resolves the real path of both the target and the root, so a
  symlink planted inside the root cannot escape it and a legitimate file reached
  through a symlinked root is not falsely refused;
- every untrusted string that reaches output -- ids, keys, paths, messages,
  history entries, raw lines and the diagnostics on stderr -- is bounded and
  escaped, not only the evidence field. The escaped set is C0, DEL, C1
  (U+0085 and U+009B among them), U+2028, U+2029 and the bidi controls
  U+200E, U+200F, U+202A-U+202E and U+2066-U+2069, each tested through a real
  run and through an identifier as well as an excerpt;
- ordering is by UTF-16 code unit, pinned by the exact list each entry point
  emits for inputs that collation orders differently -- `Z` before `a`, `README`
  before `assets`, `MAXB` before `MAX_A`, `a-b` before `a_b`. Substituting a
  collator for `byCodeUnit` fails those tests rather than passing a source scan;
- unknown configuration keys, unknown limits and unknown query filters are
  refused rather than ignored, and so is a `--since` or `--until` that is not a
  UTC instant in the one spelling `recordedAt` uses;
- an input that could not be read still produces a report naming it, for
  `history` as well as `verify`: exit 2 carries the report, and only a usage or
  configuration error leaves stdout empty.

No release has been published.


src/event.mjs
/**
 * The event record: its fields, its canonical form, and its hash.
 *
 * An event is deliberately flat. Every field is a string or null, there is no
 * nesting, and the key order of the canonical form is fixed in code rather than
 * derived from the object. That is what makes the hash reproducible on another
 * machine, in another Node build, in another locale.
 */

import { createHash } from 'node:crypto'

/** The hashed fields, in canonical order. `hash` is not one of them: it covers them. */
export const EVENT_FIELDS = Object.freeze([
  'id',
  'subject',
  'action',
  'beforeHash',
  'afterHash',
  'reason',
  'owner',
  'releaseId',
  'recordedAt',
  'corrects',
  'previousHash',
])

/** Every key a stored event line carries. Anything else is rejected, never ignored. */
export const EVENT_KEYS = Object.freeze([...EVENT_FIELDS, 'hash'])

export const ACTIONS = Object.freeze(['create', 'update', 'delete', 'correct'])

/** Fields that must be a non-empty string. The rest are a string or null. */
const REQUIRED_STRINGS = Object.freeze(['id', 'subject', 'action', 'reason', 'owner', 'releaseId', 'recordedAt', 'hash'])
const NULLABLE_STRINGS = Object.freeze(['beforeHash', 'afterHash', 'corrects', 'previousHash'])
const HASH_FIELDS = Object.freeze(['beforeHash', 'afterHash', 'previousHash', 'hash'])

export const DEFAULT_LIMITS = Object.freeze({
  maxFiles: 64,
  maxBytes: 8_000_000,
  maxEvents: 50_000,
  maxEventBytes: 16_384,
  maxFieldLength: 512,
  maxReasonLength: 1024,
  maxQueryResults: 1_000,
  timeLimitMs: 10_000,
})

const HASH_PATTERN = /^sha256:[0-9a-f]{64}$/
const TIMESTAMP_PATTERN = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/

/** A refusal to build or append an event. Never a silently dropped record. */
export class EventError extends Error {}

/**
 * Order by UTF-16 code unit, deliberately not by locale.
 *
 * `localeCompare` depends on ICU data that varies between Node builds, so two
 * correct machines can disagree about the order of the same list. Sorting by
 * code unit is the only ordering this tool ever uses.
 */
export function byCodeUnit(left, right) {
  if (left === right) return 0
  return left < right ? -1 : 1
}

/**
 * A character no field may carry: it would end or forge a line somewhere.
 *
 * C0 and DEL are the obvious ones. C1 is the range that is easy to forget and
 * just as dangerous: U+0085 is NEL, a line break to a terminal and to several
 * text readers, and U+009B is the 8-bit CSI that starts a terminal escape
 * sequence. U+2028 and U+2029 end a line for a JavaScript parser.
 */
function isControlCodePoint(code) {
  return code < 0x20
    || code === 0x7f
    || (code >= 0x80 && code <= 0x9f)
    || code === 0x2028
    || code === 0x2029
}

/**
 * Everything above, plus the bidirectional formatting characters.
 *
 * U+202E (RIGHT-TO-LEFT OVERRIDE) and its relatives do not end a line; they
 * reverse or hide what is displayed after them, which is how a crafted id makes
 * a report read as something other than what it says. They are legitimate
 * content in right-to-left text, so they are accepted inside a field and
 * escaped on the way out rather than refused on the way in.
 */
function isUnsafeInOutput(code) {
  return isControlCodePoint(code)
    || code === 0x200e
    || code === 0x200f
    || (code >= 0x202a && code <= 0x202e)
    || (code >= 0x2066 && code <= 0x2069)
}

/** True when a string carries no control character, DEL, C1, or line separator. */
export function isControlFree(text) {
  for (const character of text) {
    if (isControlCodePoint(character.codePointAt(0))) return false
  }
  return true
}

/** Write every unsafe character as its own escape text, leaving the rest exactly as it is. */
function escapeUnsafe(characters) {
  let out = ''
  for (const character of characters) {
    const code = character.codePointAt(0)
    out += isUnsafeInOutput(code) ? `\\u${code.toString(16).padStart(4, '0')}` : character
  }
  return out
}

/**
 * Bound and escape an untrusted string on its way into output.
 *
 * Every untrusted string reaches the report through this function -- ids, keys,
 * paths, messages, raw lines -- not only the evidence field. An id carrying a
 * newline or a NEL must not be able to forge an extra line in the human report,
 * a line separator must not split a JSON value in a consumer that treats U+2028
 * as a break, and a right-to-left override must not reverse the text a reader
 * sees.
 */
export function sanitize(text, max = 120) {
  const characters = Array.from(String(text))
  const bounded = characters.length > max ? [...characters.slice(0, max), ' [...]'] : characters
  return escapeUnsafe(bounded)
}

/**
 * The part of a `JSON.parse` failure that may safely be repeated.
 *
 * V8 reports a parse failure two ways, and one of them quotes the input:
 * `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid JSON`. A ledger
 * line -- or a configuration file -- short enough to be only a credential is
 * therefore reproduced in full by its own error message, and a longer one is
 * reproduced ten characters at a time, in a window around the offending
 * character. That window is drawn from wherever the error is, so it can show
 * bytes from past the 120-character bound `sanitize` applies to evidence.
 *
 * `sanitize` does not help on its own: it escapes control characters and cuts
 * from the *end*, while the quoted span sits at the front of the message.
 *
 * The quoted form carries no position, so nothing diagnostic is lost by
 * reducing it to the offending token. The other form is all position and no
 * input, and is kept. The quoted window never leaves this function.
 *
 * The quoted form is matched first on purpose: a line whose own bytes read
 * `at position 12` would otherwise be sliced after its own quoted copy.
 */
export function parseFailureDetail(error) {
  const message = typeof error?.message === 'string' ? error.message : ''
  const token = /^Unexpected token (.+?), (\.\.\.)?".*"(?:\.\.\.)? is not valid JSON$/s.exec(message)
  if (token) {
    const where = token[2] === undefined ? ' near the start' : ''
    return `unexpected token ${sanitize(token[1], 8)}${where}`
  }
  const position = /at position \d+(?: \(line \d+ column \d+\))?/.exec(message)
  if (position) return message.slice(0, position.index + position[0].length)
  if (/^Unexpected end of JSON input$/.test(message)) return message
  return 'it could not be parsed as JSON'
}

/** True when a string is a UTC instant in the one spelling this tool accepts. */
export function isUtcTimestamp(value) {
  return typeof value === 'string' && TIMESTAMP_PATTERN.test(value) && new Date(value).toISOString() === value
}

/** `sha256:<hex>` over raw bytes. The one content-hash spelling this tool writes. */
export function hashContent(bytes) {
  return `sha256:${createHash('sha256').update(bytes).digest('hex')}`
}

/** The hashed projection of an event: the declared fields, in the declared order. */
export function canonicalEvent(event) {
  const ordered = {}
  for (const field of EVENT_FIELDS) ordered[field] = event[field] ?? null
  return ordered
}

/**
 * The hash of one event, covering `previousHash`.
 *
 * Because the previous event's hash is inside this event's hash, changing any
 * event changes every hash after it. JSON.stringify escapes quotes, backslashes
 * and control characters, so no field value can imitate the framing of another.
 */
export function eventHash(event) {
  return hashContent(Buffer.from(JSON.stringify(canonicalEvent(event)), 'utf8'))
}

/**
 * One event as one line of the ledger file.
 *
 * `JSON.stringify` escapes C0 and DEL, and every other character this tool
 * refuses to display raw -- C1, the line separators and the bidi controls -- is
 * written as escape text here. Field validation already refuses the control
 * characters, so this is the second of two defences rather than the only one,
 * and an escape is the same string to any JSON reader.
 */
export function serializeEvent(event) {
  const ordered = {}
  for (const key of EVENT_KEYS) ordered[key] = event[key] ?? null
  return escapeUnsafe(JSON.stringify(ordered))
}

function fieldProblem(field, detail) {
  return { kind: 'invalid', field, detail }
}

/**
 * Check one decoded line against the event shape.
 *
 * Returns every problem it finds rather than the first, so one malformed event
 * does not hide the next. It never repairs, defaults or coerces a value: an
 * event that is not exactly what it claims to be is reported, not fixed.
 */
export function validateEventShape(value, limits = DEFAULT_LIMITS) {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    return [{ kind: 'not-object', field: null, detail: 'an event must be a JSON object' }]
  }

  const problems = []
  for (const key of Object.keys(value).sort(byCodeUnit)) {
    if (!EVENT_KEYS.includes(key)) {
      problems.push({ kind: 'unknown', field: key, detail: `unknown field "${sanitize(key, 64)}"` })
    }
  }

  for (const field of EVENT_KEYS) {
    if (!Object.hasOwn(value, field)) {
      problems.push({ kind: 'missing', field, detail: `required field "${field}" is missing` })
      continue
    }
    const raw = value[field]
    const nullable = NULLABLE_STRINGS.includes(field)
    if (raw === null) {
      if (!nullable) problems.push(fieldProblem(field, `"${field}" must be a non-empty string`))
      continue
    }
    if (typeof raw !== 'string') {
      problems.push(fieldProblem(field, `"${field}" must be a ${nullable ? 'string or null' : 'string'}`))
      continue
    }
    if (raw === '') {
      problems.push(fieldProblem(field, `"${field}" must not be empty; use null where a value does not apply`))
      continue
    }
    if (!isControlFree(raw)) {
      problems.push(fieldProblem(field, `"${field}" contains a control character or line separator`))
      continue
    }
    /* Hashes and the timestamp are bounded by their own exact formats, checked
       below. The configurable length limits govern the free-text fields only,
       so lowering one can never make a well-formed hash unrepresentable. */
    if (!HASH_FIELDS.includes(field) && field !== 'recordedAt') {
      const allowed = field === 'reason' ? limits.maxReasonLength : limits.maxFieldLength
      if (Array.from(raw).length > allowed) {
        problems.push(fieldProblem(field, `"${field}" is longer than the ${
          field === 'reason' ? 'maxReasonLength' : 'maxFieldLength'} limit of ${allowed}`))
        continue
      }
    }
    if (HASH_FIELDS.includes(field) && !HASH_PATTERN.test(raw)) {
      problems.push(fieldProblem(field, `"${field}" must look like sha256:<64 lowercase hex digits>`))
      continue
    }
    if (field === 'action' && !ACTIONS.includes(raw)) {
      problems.push(fieldProblem(field, `"action" must be one of ${ACTIONS.join(', ')}`))
      continue
    }
    if (field === 'recordedAt' && !isUtcTimestamp(raw)) {
      problems.push(fieldProblem(field, '"recordedAt" must be a UTC timestamp such as 2026-09-13T09:30:00.000Z'))
    }
  }

  if (problems.length > 0) return problems
  return actionProblems(value)
}

/** What each action promises about the before and after hashes it records. */
function actionProblems(event) {
  const problems = []
  const { action, beforeHash, afterHash, corrects } = event
  if (action === 'create') {
    if (beforeHash !== null) problems.push(fieldProblem('beforeHash', 'a "create" event must record beforeHash null'))
    if (afterHash === null) problems.push(fieldProblem('afterHash', 'a "create" event must record an afterHash'))
  }
  if (action === 'update') {
    if (beforeHash === null) problems.push(fieldProblem('beforeHash', 'an "update" event must record a beforeHash'))
    if (afterHash === null) problems.push(fieldProblem('afterHash', 'an "update" event must record an afterHash'))
  }
  if (action === 'delete') {
    if (beforeHash === null) problems.push(fieldProblem('beforeHash', 'a "delete" event must record a beforeHash'))
    if (afterHash !== null) problems.push(fieldProblem('afterHash', 'a "delete" event must record afterHash null'))
  }
  if (action === 'correct') {
    if (corrects === null) problems.push(fieldProblem('corrects', 'a "correct" event must name the event id it corrects'))
  } else if (corrects !== null) {
    problems.push(fieldProblem('corrects', '"corrects" belongs to a "correct" event; history is never rewritten in place'))
  }
  return problems
}

/**
 * Build one finished event from a draft and the hash of the event before it.
 *
 * `previousHash` is derived here and may not be supplied: a caller that could
 * choose its own predecessor could forge a chain. `recordedAt` must come from
 * the caller, because this library never reads the clock -- the command line
 * injects one, and a test injects a fixed one.
 */
export function createEvent(draft, previousHash = null, limits = DEFAULT_LIMITS) {
  if (draft === null || typeof draft !== 'object' || Array.isArray(draft)) {
    throw new EventError('An event draft must be an object')
  }
  for (const key of Object.keys(draft).sort(byCodeUnit)) {
    if (key === 'previousHash' || key === 'hash') {
      throw new EventError(`"${key}" is derived from the ledger and must not be supplied`)
    }
    if (!EVENT_FIELDS.includes(key)) throw new EventError(`Unknown event field "${sanitize(key, 64)}"`)
  }

  const event = {}
  for (const field of EVENT_FIELDS) {
    event[field] = field === 'previousHash' ? previousHash : (draft[field] ?? null)
  }
  event.hash = eventHash(event)

  const problems = validateEventShape(event, limits)
  if (problems.length > 0) throw new EventError(problems.map((problem) => problem.detail).join('; '))
  return Object.freeze(event)
}
